core/src/main/kotlin/org/evomaster/core/problem/external/service/ExternalServiceUtils.kt

```diff
@@ -31,36 +38,53 @@ object ExternalServiceUtils {
                 }
             }
         }
-        var ip = String.format("%s.%s.%s.%s", tokens[0], tokens[1], tokens[2], tokens[3])
         if (tokens[0].toInt() != 127) {
-            throw Exception("Next available IP address is out of usable range")
+            throw IllegalStateException("Next available IP address is out of usable range")
         }
-        // In the loopback address range 127.0.0.0/8, 127.255.255.255 will be the broadcast
-        // address. 127.0.0.1 is skipped because the default loopback address and used in
-        // other services commonly. 127.0.0.0 skipped because is the network address with
-        // the mask 255.0.0.0 describes the whole loopback addresses.
-        val reservedIPAddresses = arrayOf("127.0.0.0", "127.255.255.255", "127.0.0.1")
-        if (reservedIPAddresses.contains(ip)) {
+        var ip = String.format("%s.%s.%s.%s", tokens[0], tokens[1], tokens[2], tokens[3])
+        if (isReservedIP(ip)) {
             ip = nextIPAddress(ip)
         }
         return ip
     }
 
+    /**
+     * In the loopback address range 127.0.0.0/8, 127.255.255.255 will be the broadcast
+     * address. 127.0.0.1 is skipped because the default loopback address and used in
+     * other services commonly. 127.0.0.0 skipped because is the network address with
+     * the mask 255.0.0.0 describes the whole loopback addresses.
+     */
+    private fun isReservedIP(ip: String) : Boolean {
+        val reservedIPAddresses = arrayOf("127.0.0.0", "127.255.255.255", "127.0.0.1")
+        if (reservedIPAddresses.contains(ip)) {
+            return true
+        }
+        return false
+    }
+
     /**
      * This will generate random IP address for loopback range.
      *
      * Note: there is a chance for randomly generated IP address from different
      * execution to end-up closer to each other. Although the likelihood is low, but
      * can happen. As the result when creating the next IP address from the last used,
      * there will be negligible amount of small impact on the speed of the execution.
+     *
+     * In case if the generated IP address is under the reserved range, program will
+     * try to regenerate a new IP instead throwing an exception.
      */
     fun generateRandomIPAddress() : String {
         val (p1, p2, p3) = Triple(
-            (0..255).shuffled().first(),
-            (0..255).shuffled().first(),
-            (0..254).shuffled().first(),
+            randomness.randomIPBit(),
+            randomness.randomIPBit(),
+            randomness.randomIPBit(),
         )
-        return String.format("127.%s.%s.%s", p1, p2, p3)
+        var ip = String.format("127.%s.%s.%s", p1, p2, p3)
+
+        if (isReservedIP(ip)) {
```

Comment:
there is a tiny probability that this can fail second time. should be inside a loop